### PR TITLE
Free diagram

### PR DESCRIPTION
**Pull request by hmarchadour**
### Type of this PR 

- [ ] Bug fix
- [x] New feature or improvement
- [ ] Documentation
- [ ] Cleanup
- [ ] Test
- [ ] Build/Releng

### Issue(s)

#115 #116 #117 #118 #127 #129 #130 #131 #132 #133 #134 #135 #136

### What does this PR do?

A set of enhancement to obtain a "free diagram".
A "free diagram" is a diagram with unsynchronized node description. 

### Potential side effects

heavy on the diagram refresh

### How to test this PR?

- [ ] Frontend Unit tests
- [ ] Backend Unit tests
- [ ] Cypress : please specify
- [ ] Manual Test : please specify

### Checklist

- [ ] I have read CONTRIBUTING carefully.
- [ ] I have signed the [Contributor License Agreement](http://www.eclipse.org/legal/CLA.php).
- [ ] All my commits are signed-off (`-s`) with my mail address of my Eclipse Account.
- [ ] I have covered my changes by unit tests or integration tests or manual tests.
- [ ] All tests pass.
- [ ] I have updated the documentation accordingly
- [ ] New React components are availables in storybook

### CODE REVIEW

File: frontend/src/diagram/DiagramWebSocketContainer.tsx [hunk 1]
```diff
@@ -229,15 +231,56 @@ export const DiagramWebSocketContainer = ({ representationId, selection, setSele
...
   const [getToolSectionData, { loading: toolSectionLoading, data: toolSectionData }] = useLazyQuery(
     getToolSectionsQuery
   );
+
+  const setErrorMessage = useCallback((errorMessage?) => {
+    dispatch({ type: HANDLE_ERROR_MESSAGE__ACTION, errorMessage });
+  }, []);
+
+  useEffect(() => {
+    if (!deleteElementResult?.loading && deleteElementResult?.data?.data) {
```
Comment (pcdavid) [inline frontend/src/diagram/DiagramWebSocketContainer.tsx:257]: I beliave the double `data` (here and below) is wrong now that we use Apollo Client.

**Review by pcdavid — CHANGES_REQUESTED** [review]

File: backend/sirius-web-collaborative-diagrams-api/src/main/java/org/eclipse/sirius/web/collaborative/diagrams/api/IDiagramService.java [hunk 2]
```diff
@@ -27,6 +27,8 @@
 public interface IDiagramService {
     Diagram create(DiagramCreationParameters parameters);
 
+    Diagram create(DiagramCreationParameters parameters, Diagram previousDiagram);
```
Comment (pcdavid) [inline backend/sirius-web-collaborative-diagrams-api/src/main/java/org/eclipse/sirius/web/collaborative/diagrams/api/IDiagramService.java:30]: This seems... strange. We ask for the _creation_ (not update or refresh) of a diagram, but give an already `previousDiagram` as a parameter? Maybe it's juste the method and/or parameter name that should be different/more explicit, but in terms of API its confusing.

File: backend/sirius-web-spring-collaborative-diagrams/src/main/java/org/eclipse/sirius/web/spring/collaborative/diagrams/DiagramService.java [hunk 3]
```diff
@@ -109,4 +137,45 @@ public Diagram create(DiagramCreationParameters parameters) {
...
+        return result;
+    }
+
+    private List<String> getTargetIds(Node node) {
+        List<String> result = new ArrayList<>();
+        result.add(node.getTargetObjectId());
+        node.getBorderNodes().stream().map(this::getTargetIds).flatMap(List::stream).forEach(result::add);
+        node.getChildNodes().stream().map(this::getTargetIds).flatMap(List::stream).forEach(result::add);
+        return result;
+    }
```
Comment (pcdavid) [inline backend/sirius-web-spring-collaborative-diagrams/src/main/java/org/eclipse/sirius/web/spring/collaborative/diagrams/DiagramService.java:180]: This seems to allocate quite a lot of intermediate lists.

 How about something like (untested):

```java
    private List<String> getTargetIds(Diagram diagram) {
        List<String> result = new ArrayList<>();
        result.add(diagram.getTargetObjectId());
        diagram.getNodes().forEach(node -> this.collectTargetIds(node, result));
        return result;
    }

    private void collectTargetIds(Node node, List<String> result) {
        result.add(node.getTargetObjectId());
        node.getBorderNodes().stream().map(borderNode -> this.collectTargetIds(borderNode, result));
        node.getChildNodes().stream().map(childNode -> this.collectTargetIds(childNode, result));
    }
```

Side-note:  maybe `Diagram` should have something like `forEachNode(Consumer<Node>)`.

File: backend/sirius-web-spring-collaborative-diagrams/src/main/java/org/eclipse/sirius/web/spring/collaborative/diagrams/DiagramContext.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+import org.eclipse.sirius.web.collaborative.diagrams.api.IDiagramContext;
+import org.eclipse.sirius.web.diagrams.Diagram;
+
+/**
+ * The implementation of {@link IDiagramContext}.
+ *
+ * @author hmarchadour
+ */
+public class DiagramContext implements IDiagramContext {
```
Comment (pcdavid) [inline backend/sirius-web-spring-collaborative-diagrams/src/main/java/org/eclipse/sirius/web/spring/collaborative/diagrams/DiagramContext.java:25]: This feels like reinventing something like `AtomicReference` (without the thread safety).

File: backend/sirius-web-diagrams/src/main/java/org/eclipse/sirius/web/diagrams/components/DiagramComponent.java [hunk 5]
```diff
@@ -52,9 +55,18 @@ public Element render() {
         DiagramRenderingCache cache = new DiagramRenderingCache();
 
         // @formatter:off
+        List<Node> allPrevNodes = this.props.getPreviousDiagram()
+                .map(Diagram::getNodes)
```
Comment (pcdavid) [inline backend/sirius-web-diagrams/src/main/java/org/eclipse/sirius/web/diagrams/components/DiagramComponent.java:59]: This is only the top-level nodes. Is it expected? If yes, maybe `topLevelNodes` would be a better name.

**Review by pcdavid — CHANGES_REQUESTED** [review]

File: backend/sirius-web-diagrams/src/main/java/org/eclipse/sirius/web/diagrams/components/NodeComponent.java [hunk 6]
```diff
@@ -47,72 +51,104 @@ public NodeComponent(NodeComponentProps props) {
...
-            cache.put(semanticElement, nodeElement);
+        } else {
+            Optional<Map> optionalTargetSemanticObjects = variableManager.get(DiagramDescription.TARGET_SEMANTIC_OBJECTS, Map.class);
+            if (optionalTargetSemanticObjects.isPresent()) {
+                Map<String, Object> targetSemanticObjects = optionalTargetSemanticObjects.get();
+               // @formatter:off
+               this.props.getPreviousNodes()
+                   .forEach(prevNode -> {
+                       String targetObjectId = prevNode.getTargetObjectId();
+                       if (targetSemanticObjects.containsKey(targetObjectId)) {
```
Comment (sbegaudeau) [inline backend/sirius-web-diagrams/src/main/java/org/eclipse/sirius/web/diagrams/components/NodeComponent.java:68]: You are making sure that the targetId from a previous node is contained in TARGET_SEMANTIC_OBJECTS which has been created just a couple methods calls ago using the same previous node. I fail to grasp the meaning of what you are doing and why you are computing this information this way.

Starting with an ADR to properly define the problem and the solution may be helpful.

File: backend/sirius-web-spring-collaborative-diagrams/src/main/java/org/eclipse/sirius/web/spring/collaborative/diagrams/DiagramService.java [hunk 7]
```diff
@@ -48,29 +53,52 @@
...
 
-        DiagramComponentProps props = new DiagramComponentProps(variableManager, parameters.getDiagramDescription());
+        DiagramComponentProps props;
+        if (optionalPreviousDiagram.isPresent()) {
+            Diagram previousDiagram = optionalPreviousDiagram.get();
+            if (this.containsUnsyncDescriptions(diagramDescription)) {
+                Map<String, Object> targetSemanticObjectsMap = this.getSemanticElements(previousDiagram, parameters.getEditingContext());
+                variableManager.put(DiagramDescription.TARGET_SEMANTIC_OBJECTS, targetSemanticObjectsMap);
+            }
+            props = new DiagramComponentProps(variableManager, diagramDescription, previousDiagram);
```
Comment (sbegaudeau) [inline backend/sirius-web-spring-collaborative-diagrams/src/main/java/org/eclipse/sirius/web/spring/collaborative/diagrams/DiagramService.java:96]: There are two ways to give information to a component, using the props directly or thanks to the variable manager. Both ways have specific meanings and the variable manager should not be considered as "the way to give something recursively to every components". It's only data which will be used by the various providers of the descriptions such as the semantic candidates provider or the target object id provider.

In other words, adding something to the variable manager means adding a new public API that specifiers can rely on to create their providers and expressions if they rely on the compatibility layer. The `targetSemanticObjectsMap` should not be visible to specifiers.

File: backend/sirius-web-collaborative-diagrams-api/src/main/java/org/eclipse/sirius/web/collaborative/diagrams/api/IDiagramService.java (hunk 2 above)
Comment (sbegaudeau) [reply backend/sirius-web-collaborative-diagrams-api/src/main/java/org/eclipse/sirius/web/collaborative/diagrams/api/IDiagramService.java:30]: I have rewrote that

File: backend/sirius-web-diagrams/src/main/java/org/eclipse/sirius/web/diagrams/components/NodeComponent.java [hunk 8]
```diff
@@ -47,72 +51,104 @@ public NodeComponent(NodeComponentProps props) {
...
-            children.add(nodeElement);
-
-            cache.put(nodeDescription.getId(), nodeElement);
-            cache.put(semanticElement, nodeElement);
+        } else {
+            Optional<Map> optionalTargetSemanticObjects = variableManager.get(DiagramDescription.TARGET_SEMANTIC_OBJECTS, Map.class);
+            if (optionalTargetSemanticObjects.isPresent()) {
+                Map<String, Object> targetSemanticObjects = optionalTargetSemanticObjects.get();
+               // @formatter:off
+               this.props.getPreviousNodes()
```
Comment (sbegaudeau) [inline backend/sirius-web-diagrams/src/main/java/org/eclipse/sirius/web/diagrams/components/NodeComponent.java:65]: I am not sure that I understand your algorithm. With your strategy, if two semantic elements had their order changed, you would still render them with the old order.

**Comment by sbegaudeau** [comment]
I'll close this PR for now because it cannot be merged since:

- Some parts have already been merged
- Other parts need to be updated / rewritten / etc
- Other parts are now completely outdated

It may very well be reopened in the future or new PRs may be created from updated versions of some commits in the future but this one will never be merged as is.
